bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go

```diff
@@ -95,6 +98,31 @@ type KubeadmConfigSpec struct {
 	// For more information, refer to https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055.
 	// +optional
 	UseExperimentalRetryJoin bool `json:"useExperimentalRetryJoin,omitempty"`
+
+	// IgnitionConfig contains Ignition specific configuration.
+	//
+	// If set, bootstrap data will be generated in Ignition format instead of cloud-init.
+	//
+	// +optional
+	IgnitionConfig *IgnitionConfig `json:"ignitionConfig,omitempty"`
+}
+
+// IgnitionConfig contains Ignition specific configuration.
+type IgnitionConfig struct {
+	// ContainerLinuxConfig contains CLC specific configuration.
+	//
+	// +optional
+	ContainerLinuxConfig *ContainerLinuxConfig `json:"containerLinuxConfig,omitempty"`
+}
+
+// ContainerLinuxConfig contains CLC specific configuration.
+//
+// We use structured type here to keep space for adding additional fields, for example 'version'.
+type ContainerLinuxConfig struct {
```

Comment:
```suggestion
type ContainerLinuxSpec struct {
```

Reply:
Let's discuss this in https://github.com/kubernetes-sigs/cluster-api/pull/4172#discussion_r578555137.